### PR TITLE
Fix job cancel chain bugs

### PR DESCRIPTION
**Pull request by AlanCoding**
##### SUMMARY
I have been hitting a mysterious bug in the collection tests (running here) specifically. However, I still believed it to be very very legitimate.

Several of these tests seem to have confirmed the issue.

I believe I saw it manually with bulk jobs, cc @kdelee 
 - go to the /bulk_launch/ endpoint
 - launch a bunch of "Demo Job Template" runs
 - then try to cancel the workflow job

I believe this same bug manifested in that case because all of the jobs are blocked on the first job that started running, which then went to the canceled state.

##### ISSUE TYPE
 - Bug, Docs Fix or other nominal change

##### COMPONENT NAME
 - API

---

> [!NOTE]
> **Medium Risk**
> Touches core scheduling/cancellation paths in `TaskManager`/`WorkflowManager` and workflow DAG cancellation logic; mistakes could leave jobs or workflows stuck non-terminal or mis-labeled, but changes are targeted and covered by new functional tests.
> 
> **Overview**
> Fixes several cancellation edge cases so workflow and dependency chains reach terminal states promptly and consistently.
> 
> `WorkflowApproval.cancel()` now immediately transitions approvals to `canceled` (since they are excluded from TaskManager processing). Workflow cancellation handling now only considers cancel complete when child jobs are actually terminal, and it proactively reschedules `TaskManager`/`WorkflowManager` to speed up cancel propagation.
> 
> `TaskManager` now (a) cancels *pending* jobs with `cancel_flag` directly to `canceled`, and (b) treats *canceled* dependencies like failed ones, with a distinct `job_explanation`. Adds a new functional test suite covering canceled dependencies, pending-job cancel handling, and workflow cancel finalization (including approval nodes).
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 2df1ec0e3a37b034e758e1717feec03deff58357. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

## Summary by CodeRabbit

* **Bug Fixes**
  * Pending jobs with cancel flags are now immediately canceled without waiting for the next task cycle.
  * Canceled dependencies now properly fail dependent jobs with appropriate explanations.
  * WorkflowApproval cancellations are processed immediately.
  * Workflow finalization is accelerated after processing canceled or failed tasks.

* **Tests**
  * Added comprehensive tests for cancellation propagation across dependencies and workflows.
  * Added performance benchmark test for job execution.

### CODE REVIEW

File: awx/main/models/workflow.py [hunk 1]
```diff
@@ -918,6 +918,17 @@ def deny(self, request=None):
...
+    def cancel(self, job_explanation=None, is_chain=False):
+        # WorkflowApprovals have no dispatcher process (they wait for human
+        # input) and are excluded from TaskManager processing, so the base
+        # cancel() would only set cancel_flag without ever transitioning the
+        # status.  We call super() for the flag, then transition directly.
+        super().cancel(job_explanation=job_explanation, is_chain=is_chain)
+        if self.status != 'canceled':
+            self.status = 'canceled'
+            self.save(update_fields=['status'])
+            self.websocket_emit_status('canceled')
```
[the remark this answers is not in the document]
Comment (AlanCoding) [reply awx/main/models/workflow.py]: You're wrong. Because in this same diff, `cancel_finished` local variable in the workfow manager is not set to `False` in the case when the status of the job is "canceled". So that should be picking up this save, and if the only job active in the workflow is approval nodes, then the WorkflowJob will go straight to canceled state. And if some other job type needs rescheduling, that's out of scope here.

[the remark this answers is not in the document]
Comment (AlanCoding) [reply awx/main/models/workflow.py]: valid enough. Will push a commit.

File: awx/main/scheduler/task_manager.py [hunk 2]
```diff
@@ -538,15 +551,24 @@ def process_running_tasks(self, running_tasks):
...
                 break
 
-            has_failed = self.process_job_dep_failures(task)
-            if has_failed:
+            if task.cancel_flag:
+                task.status = 'canceled'
+                task.job_explanation = gettext_noop("This job was canceled before it started.")
+                tasks_to_cancel.append(task)
+                logger.debug(f"Canceling pending task {task.log_format} because cancel_flag is set")
+                continue
```
[the remark this answers is not in the document]
Comment (AlanCoding) [reply awx/main/scheduler/task_manager.py:572]: Someone can try to come in and correct me, but I'll bet money that `start_task_limit` is designed to _process_ that number of pending jobs. It's not just to "look" at the number of jobs.

Because _any_ implementation that just "looks" at that number and quits is subject to a deadlock type of state, where it never makes any progress because the first (ordered) 100 jobs don't go anywhere.

Given that, there's no reason for job cancels to "jump the line" in this respect. It is already a duty to ensure that the task manager makes steady progress moving jobs onto the next status. And transitioning the status to "canceled" isn't fundamentally different from the other work that it's trying to do (like placing jobs on nodes). So canceled pending jobs should wait in line for the task manager just like all the other jobs. Otherwise you're asking to write a special case for this, which is ultimately a corner case. That is absolutely opposite of best practice and this should be rejected. I think the comment here only makes sense believing that we have a no-progress risk, and that's a totally separate subject and already needs to have guardrails, and we expect it does. This is unlike the _task manager timeout_ which really **does have a serious** "no-progress" risk. And I changed my implementation because of that, because processing cancel `status` change needs to happen in the commit in those cases, otherwise it can deadlock. Risks of task manager timeout are **heavily correlated** with a large pending job cancel volume, and system stability is best maintained by doing the `.save(update_fields` right away when the circumstance is seen.

File: awx/main/scheduler/task_manager.py [hunk 3]
```diff
@@ -545,8 +561,17 @@ def process_pending_tasks(self, pending_tasks):
...
                 break
 
-            has_failed = self.process_job_dep_failures(task)
-            if has_failed:
+            if task.cancel_flag:
+                logger.debug(f"Canceling pending task {task.log_format} because cancel_flag is set")
+                task.status = 'canceled'
+                task.job_explanation = gettext_noop("This job was canceled before it started.")
+                task.save(update_fields=['status', 'job_explanation'])
+                task.websocket_emit_status('failed')
```
[the remark this answers is not in the document]
Comment (AlanCoding) [reply awx/main/scheduler/task_manager.py]: implemented

**Review by fosterseth — APPROVED** [review]

File: awx/main/scheduler/task_manager.py [hunk 4]
```diff
@@ -443,19 +447,32 @@ def after_lock_init(self):
...
-                )
+                if dep.status == 'canceled':
+                    logger.warning(f'Previous task canceled, failing task: {task.id} dep: {dep.id} task manager')
+                    task.job_explanation = 'Previous Task Canceled: {"job_type": "%s", "job_name": "%s", "job_id": "%s"}' % (
+                        get_type_for_model(type(dep)),
+                        dep.name,
+                        dep.id,
+                    )
+                    ScheduleWorkflowManager().schedule()  # speedup for dependency chains in workflow, on workflow cancel
+                    task.websocket_emit_status('canceled')
```
[the remark this answers is not in the document]
Comment (AlanCoding) [reply awx/main/scheduler/task_manager.py]: ok fine, will push commit

File: awx/main/scheduler/dag_workflow.py [hunk 5]
```diff
@@ -122,8 +123,15 @@ def cancel_node_jobs(self):
...
                 continue
             elif job.can_cancel:
-                cancel_finished = False
                 job.cancel()
+                # WorkflowApprovals are excluded from TaskManager processing,
+                # so we must transition them directly here.
+                if isinstance(job.get_real_instance(), WorkflowApproval):
+                    job.status = 'canceled'
+                    job.save(update_fields=['status'])
+                    job.websocket_emit_status('canceled')
```
[the remark this answers is not in the document]
Comment (AlanCoding) [reply awx/main/scheduler/dag_workflow.py]: Only called by task manager which has authority to change status, so not a problem
